wire-runtime/src/main/java/com/squareup/wire/Message.java

```diff
@@ -160,46 +161,40 @@ protected void setBuilder(Builder builder) {
     return adapter.fromInt(value);
   }
 
+  /**
+   * Create a {@code byte[]} of the serialized form of the data.
+   * <p>
+   * Note: This is an inefficient means of serializing data. For best performance, call
+   * {@link Wire#adapter} and then {@link MessageAdapter2#write}.
+   */
   @SuppressWarnings("unchecked")
   public byte[] toByteArray() {
-    return WIRE.messageAdapter((Class<Message>) getClass()).toByteArray(this);
-  }
-
-  public void writeTo(byte[] output) {
-    writeTo(output, 0, output.length);
-  }
-
-  public void writeTo(byte[] output, int offset, int count) {
-    write(WireOutput.newInstance(output, offset, count));
-  }
-
-  @SuppressWarnings("unchecked")
-  private void write(WireOutput output) {
-    MessageAdapter<Message> adapter = WIRE.messageAdapter((Class<Message>) getClass());
+    Buffer buffer = new Buffer();
     try {
-      adapter.write(this, output);
+      WIRE.messageAdapter((Class<Message>) getClass()).write(this, buffer);
     } catch (IOException e) {
-      throw new RuntimeException(e);
+      throw new AssertionError(e); // No I/O writing into Buffer.
     }
+    return buffer.readByteArray();
   }
 
-  public void writeUnknownFieldMap(WireOutput output) throws IOException {
+  void writeUnknownFieldMap(WireOutput output) throws IOException {
     if (unknownFields != null) {
       unknownFields.write(output);
     }
   }
 
   @SuppressWarnings("unchecked")
-  public int getSerializedSize() {
+  int getSerializedSize() {
     if (!haveCachedSerializedSize) {
-      MessageAdapter<Message> adapter = WIRE.messageAdapter((Class<Message>) getClass());
+      WireMessageAdapter<Message> adapter = WIRE.messageAdapter((Class<Message>) getClass());
       cachedSerializedSize = adapter.getSerializedSize(this);
       haveCachedSerializedSize = true;
     }
     return cachedSerializedSize;
```

Comment:
There's a two-phase write in which this helps. First the size is determined (and written) and then the actual data is written. Would be nice to encapsulate it as local state rather than instance state in the future.